### ISSUE TITLE
Exception Spam from Windows Mixed Reality Device Manager

### ISSUE THREAD

**Issue by michael-house** [microsoft/MixedRealityToolkit-Unity#7698]
## Describe the bug

Every frame, a null pointer exception is occurring:

> NullReferenceException: Object reference not set to an instance of an object
> Microsoft.Windows.Perception.Spatial.ISpatialCoordinateSystem.op_Implicit (Microsoft.Windows.Perception.Spatial.SpatialCoordinateSystem obj) (at <ed1f6dedc2514473bacf0986f8922b29>:0)
> Microsoft.Windows.UI.Input.Spatial.SpatialPointerPose.TryGetAtTimestamp (Microsoft.Windows.Perception.Spatial.SpatialCoordinateSystem coordinateSystem, Microsoft.Windows.Perception.PerceptionTimestamp timestamp) (at <ed1f6dedc2514473bacf0986f8922b29>:0)
> Microsoft.MixedReality.Toolkit.WindowsMixedReality.Input.WindowsMixedRealityDeviceManager.Update () (at Assets/MRTK/Providers/WindowsMixedReality/XR2018/WindowsMixedRealityDeviceManager.cs:446)
> Microsoft.MixedReality.Toolkit.BaseDataProviderAccessCoreSystem.Update () (at Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:55)
> Microsoft.MixedReality.Toolkit.MixedRealityToolkit+<>c.<UpdateAllServices>b__64_0 (Microsoft.MixedReality.Toolkit.IMixedRealityService service) (at Assets/MRTK/Core/Services/MixedRealityToolkit.cs:946)
> Microsoft.MixedReality.Toolkit.MixedRealityToolkit.ExecuteOnAllServicesInOrder (System.Action`1[T] execute) (at Assets/MRTK/Core/Services/MixedRealityToolkit.cs:1034)
> Microsoft.MixedReality.Toolkit.MixedRealityToolkit.UpdateAllServices () (at Assets/MRTK/Core/Services/MixedRealityToolkit.cs:946)
> Microsoft.MixedReality.Toolkit.MixedRealityToolkit.Update () (at Assets/MRTK/Core/Services/MixedRealityToolkit.cs:630)

## To reproduce

- New Unity Project
- Bring in Assets/MRTK/ from fresh clone of mrtk_development @ https://github.com/microsoft/MixedRealityToolkit-Unity/commit/76922618bb14f59534ef3748f6088779f4b6a6b4 (or I went back a few weeks to see if I could find a build that worked and for at least the last two weeks this issue was there.)
- Switch to UWP
- Ensure MS Build is enabled

## Expected behavior

No exception.

## Your setup (please complete the following information)

- Unity Version 2018.4.17f1
- MRTK Version v2.4 (dev)

## Target platform (please complete the following information)

- ~~HoloLens 2~~
- In Unity Editor

## Additional context

It's not clear what's causing this, but I somehow managed to get it to go away. Where usually `Microsoft.Windows.UI.Input.Spatial.SpatialPointerPose.TryGetAtTimestamp (Microsoft.Windows.Perception.Spatial.SpatialCoordinateSystem coordinateSystem, Microsoft.Windows.Perception.PerceptionTimestamp timestamp)` ([here](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/76922618bb14f59534ef3748f6088779f4b6a6b4/Assets/MRTK/Providers/WindowsMixedReality/XR2018/WindowsMixedRealityDeviceManager.cs#L446)) is totally fine with you passing in a null pointer for the `coordinateSystem` (as happens in the Editor), it seems that changing some settings in the project (maybe something related to WinRT, or MSBuild?), `TryGetAtTimestamp` starts throwing an exception when a null pointer is passed in.

This bug can be avoided by checking for null on the `coordinateSystem` before calling the method (and not calling the method if it's null), but I suspect this is the tip of the iceberg for some other set of problems.

**Comment by wischie** [comment]
I'm experiencing this same issue. Not sure how to get around it, can't check out any demos or examples just keep getting spammed with this error.

**Comment by keveleigh** [comment]
Oh weird! Yeah, we should definitely be checking null here before passing it anywhere.
Any ideas why you might be getting a null `SpatialCoordinateSystem` on-device? That seems like a scenario that shouldn't ever happen.

**Comment by michael-house** [comment]
Sorry I wasn't clear about this, this is in the editor. As far as I know, this isn't an issue on device, but I haven't tested it yet.

**Comment by davidkline-ms** [comment]
When looking into this, i am seeing this only if the DotNetWinRT assembly is present when running in the editor. Haven't looked deeper than confirming a null spatial coordinate system in editor with and a non-null without.

The test i was running is using the .net backend. Could that be causing the trouble? I have not had a chance to check IL2CPP yet.

```c#
#if ENABLE_DOTNET
    spatialCoordinateSystem = GetSpatialCoordinateSystem(newSpatialCoordinateSystemPtr);
#elif WINDOWS_UWP
    spatialCoordinateSystem = Marshal.GetObjectForIUnknown(newSpatialCoordinateSystemPtr) as SpatialCoordinateSystem;
#elif DOTNETWINRT_PRESENT
    spatialCoordinateSystem = SpatialCoordinateSystem.FromNativePtr(newSpatialCoordinateSystemPtr);
#endif
```

**Comment by michael-house** [comment]
@davidkline-ms Yes, I believe that's the same I'm seeing.

The code you posted [doesn't include the conditional](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/76922618bb14f59534ef3748f6088779f4b6a6b4/Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs#L78). 

Where, I'm pretty sure `newSpatialCoordinateSystemPtr` is zero at that point. It's set by [this](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MRTK/Providers/WindowsMixedReality/XR2018/WindowsMixedRealityUtilitiesProvider.cs#L19).

And, `WorldManager.GetNativeISpatialCoordinateSystemPtr()` appears to return zero when in the Editor.

So, we end up returning `spatialCoordinateSystem` at it's default value, which is `null`. Perhaps on [line 90](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MRTK/Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs#L90) we should return `currentSpatialCoordinateSystemPtr` instead, since it has it's default value set to zero.

**Comment by davidkline-ms** [comment]
@michael-house, i probably should have included more of the code :)  the intent was to look at the preprocessor conditions to see if there was anything there

**Comment by keveleigh** [comment]
Fixed via #7710
